### PR TITLE
TLS 1.3: PSK: Add parser of psk kex mode ext on server side

### LINKED ISSUE #5561: Upstream TLS 1.3 Pre-Shared Key Exchange Modes extension parsing on server side
This issue tracks the up-streaming of the server Pre-Shared Key Exchange Modes extension parsing.

Source: https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype

Precise acceptance criteria:

- test (unit test) with an Mbed TLS client that a ClientHello containing a Pre-Shared Key Exchange Modes extension is successfully parsed (depends on #5560).
- test (ssl-opt.sh test) with an OpenSSL/GnuTLS client that a ClientHello containing a Pre-Shared Key Exchange Modes extension is successfully parsed by an Mbed TLS TLS 1.3 server.

### LINKED ISSUE #5661: Segmentation fault (core dumped) for mbedtls_ctr_drbg_random in "mbedtls/blob/development/library/ctr_drbg.c"
### Summary
use function mbedtls_ctr_drbg_random and get Segmentation fault.

### Actual behavior
Segmentation fault (core dumped)

### Steps to reproduce
--> mbedtls_ecp_gen_key
--> mbedtls_ecp_gen_keypair
--> mbedtls_ecp_gen_keypair_base-->mbedtls_ecp_gen_privkey
--> mbedtls_ctr_drbg_random
--> ret = mbedtls_ctr_drbg_random_with_add( ctx, output, output_len, NULL, 0 );

### Additional information
ret = mbedtls_ctr_drbg_random_with_add( ctx, output, output_len, NULL, 0 ); ---> will get Segmentation fault
ret = crypto_ctr_drbg_random_with_add( ctx, output, output_len, "xx", 2 ); ---> will be executed passed, but I don't know if the result is right or wrong;

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description
fix #5561

## Status
**READY**

### CODE REVIEW

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me. I have only minor comments and requests for improvement.

File: tests/ssl-opt.sh [hunk 1]
```diff
@@ -878,6 +878,8 @@ wait_client_done() {
     CLI_EXIT=$?
 
     kill $DOG_PID >/dev/null 2>&1
+    # For ubuntu 22.04, `Terminated` message is outputed from `wait` command.
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:881]:
```suggestion
    # For Ubuntu 22.04, `Terminated` message is outputted by the wait command.
```

File: tests/ssl-opt.sh [hunk 2]
```diff
@@ -878,6 +878,8 @@ wait_client_done() {
...
+    # to eliminate it from stdout, redirect stdout/stderr to CLI_OUT
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:882]:
```suggestion
    # To remove it from stdout, redirect stdout/stderr to CLI_OUT
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Just some wording to improve I think.

File: library/ssl_tls13_server.c [hunk 3]
```diff
@@ -45,6 +45,60 @@
...
+ */
+static int ssl_tls13_parse_key_exchange_modes_ext( mbedtls_ssl_context *ssl,
+                                                   const unsigned char *buf,
+                                                   const unsigned char *end )
+{
+    const unsigned char *p = buf;
+    size_t ke_modes_len;
+    int ke_modes = 0;
+
+    /* Read PSK mode list length (1 Byte) */
```
Comment (xkqian) [inline library/ssl_tls13_server.c:64]:
```suggestion
    /* Read ke_modes length (1 Byte) */
```

**Review by ronald-cron-arm — APPROVED** [review]
LGTM

**Comment by ronald-cron-arm** [comment]
@xkqian please have another look to the last version.

**Pull request merged**